File: algorithm/programmers/LV2/python/전력망을 둘로 나누기.py
from collections import deque


def solution(n, wires):
    answer = float('inf')

    connect = [True for _ in range(n-1)]

    for i in range(n-1):
        # 연결을 끊고 둘 중 한 노드를 시작으로 BFS
        connect[i] = False
        result = bfs(wires[i][1], connect, wires, n)
        answer = min(abs(n - 2 * len(result)), answer)
        connect[i] = True

    return answer


def bfs(start, connect, wires, n):
    result = [start]
    visited = [False for _ in range(n+1)]
    q = deque()
    q.append(start)

    while q:
        node = q.popleft()
        visited[node] = True

        for i in range(len(wires)):
            # 연결되어 있고 노드가 포함되어 있다면?
            if connect[i] and node in wires[i]:
                # 들린 적이 없다면 deque와 result에 추가
                for element in wires[i]:
                    if not visited[element]:
                        q.append(element)
                        result.append(element)
    return result

File: algorithm_java/src/main/java/programmers/lv2/전력망을_둘로_나누기.java
package programmers.lv2;

import java.util.*;
import java.util.stream.Collectors;

public class 전력망을_둘로_나누기 {
    int answer = Integer.MAX_VALUE;
    boolean[] connected;

    public int solution(int n, int[][] wires) {
        connected = new boolean[n];
        Arrays.fill(connected, true);

        for (int i = 0; i < n-1; i++) {
            connected[i] = false;
            bfs(wires[i][1], n, wires);
            connected[i] = true;
        }

        return answer;
    }

    public void bfs(int start, int n, int[][] wires) {
        LinkedList<Integer> queue = new LinkedList<>();
        boolean[] visited = new boolean[n + 1];
        int count = 1;
        queue.add(start);

        while (!queue.isEmpty()) {
            int now = queue.poll();
            visited[now] = true;

            for (int i = 0; i < wires.length; i++) {
                if (!connected[i]) continue;

                // int[] -> ArrayList<Integer>
                List <Integer> temp = Arrays.stream(wires[i]).boxed().collect(Collectors.toList());
                if (temp.contains(now)) {
                    for (int k = 0; k < temp.size(); k++) {
                        if (!visited[temp.get(k)] && !queue.contains(temp.get(k))) {
                            queue.add(temp.get(k));
                            count += 1;
                        }
                    }
                }
            }
        }

        answer = Math.min(Math.abs(count * 2 - n), answer);

    }
}
